drivers/usb/device/usb_dc_nrf5.c

```diff
@@ -0,0 +1,2256 @@
+		return -EALREADY;
+	}
+
+	nrf5_usbd_cfg_ep_interrupt(ep, false);
+	nrf_usbd_ep_disable(ep);
+	ep_ctx->cfg.en = false;
+
+	return 0;
+}
+
+int usb_dc_ep_flush(const u8_t ep)
+{
+	struct nrf5_usbd_ep_ctx *ep_ctx;
+
+	if (!nrf5_usbd_dev_attached() || !nrf5_usbd_dev_ready()) {
+		return -ENODEV;
+	}
+
+	ep_ctx = nrf5_usbd_endpoint_ctx(ep);
+	if (!ep_ctx) {
+		return -EINVAL;
+	}
+
+	ep_ctx->buf.len = 0;
+	ep_ctx->buf.curr = ep_ctx->buf.data;
+
+	return 0;
+}
+
+int usb_dc_ep_write(const u8_t ep, const u8_t *const data,
+		    const u32_t data_len, u32_t * const ret_bytes)
+{
+	struct nrf5_usbd_ctx *ctx = get_usbd_ctx();
+	struct nrf5_usbd_ep_ctx *ep_ctx;
+	u32_t bytes_to_copy;
+
+	if (!nrf5_usbd_dev_attached() || !nrf5_usbd_dev_ready()) {
+		return -ENODEV;
+	}
+
+	if (NRF_USBD_EPOUT_CHECK(ep)) {
+		return -EINVAL;
+	}
+
+	ep_ctx = nrf5_usbd_endpoint_ctx(ep);
+	if (!ep_ctx) {
+		return -EINVAL;
+	}
+
+	/**
+	 * TODO:
+	 * With usb_write called multiple times over before the previous
+	 * write is complete, there are data corruptions on the bus,
+	 * possibly because software over-writes the HW's local buffer
+	 * and initiates the next write before or while an ongoing
+	 * USB write on the bus.
+	 */
+	if (IS_SET(ep_ctx->flags, EP_WRITE_PENDING)) {
+		return -EAGAIN;
+	}
+
+	k_sem_take(&ctx->dma_in_use, K_FOREVER);
+
+	bytes_to_copy = min(data_len, ep_ctx->cfg.max_sz);
+	memcpy(ep_ctx->buf.data, data, bytes_to_copy);
+	ep_ctx->buf.len = bytes_to_copy;
+
+	if (ret_bytes) {
+		*ret_bytes = bytes_to_copy;
+	}
+
+	nrf_usbd_ep_easydma_set(ep, (u32_t)ep_ctx->buf.data, ep_ctx->buf.len);
+
+	switch (ep_ctx->cfg.type) {
+	case USB_DC_EP_CONTROL:
+	{
+		if (bytes_to_copy) {
+			nrf5_usbd_start_epin_task(ep);
+		} else {
+			if (get_usbd_ctx()->address_set) {
+				get_usbd_ctx()->address_set = false;
+				/**
+				 * TODO: For SET_ADDRESS, HW takes care of
+				 * initiating STATUS but for some reason
+				 * it looks like we need to start ep0status
+				 * task. Clarify this!
+				 */
+				/* break; */
+			}
+			nrf5_usbd_start_ep0status_task();
+			k_sem_give(&ctx->dma_in_use);
+			ep_ctx->state = EP_IDLE;
+		}
+	}
+		break;
+	case USB_DC_EP_BULK:
+	case USB_DC_EP_INTERRUPT:
+		/* TODO: Should we set EP_WRITE_PENDING here? */
+		nrf5_usbd_start_epin_task(ep);
+		break;
+	case USB_DC_EP_ISOCHRONOUS:
+		nrf5_usbd_start_epin_task(ep);
+		break;
+	default:
+		SYS_LOG_ERR("Invalid endpoint type");
+		break;
+	}
+
+	return 0;
+}
+
+int usb_dc_ep_read_wait(u8_t ep, u8_t *data, u32_t max_data_len,
+			u32_t *read_bytes)
+{
+	struct nrf5_usbd_ep_ctx *ep_ctx;
+	u32_t bytes_to_copy;
+
+	if (!nrf5_usbd_dev_attached() || !nrf5_usbd_dev_ready()) {
+		return -ENODEV;
+	}
+
+	if (NRF_USBD_EPIN_CHECK(ep)) {
+		return -EINVAL;
+	}
+
+	if (!data && max_data_len) {
+		return -EINVAL;
+	}
+
+	ep_ctx = nrf5_usbd_endpoint_ctx(ep);
+	if (!ep_ctx) {
+		return -EINVAL;
+	}
+
+	if (!data && !max_data_len) {
+		if (read_bytes) {
+			*read_bytes = ep_ctx->buf.len;
+		}
+		goto done;
```

Comment:
`return ret` here directly, no need for `goto`

Reply:
Ack